include/up-cpp/utils/ThreadPool.h

```diff
@@ -0,0 +1,150 @@
+                    funcPtr();
+                } 
+            }
+
+            // Submit a function to be executed asynchronously by the pool
+            template<typename F, typename...Args>
+            auto submit(F&& f, Args&&... args) -> std::future<decltype(f(args...))> {
+
+                if (true == terminate_) {
+                    spdlog::error("Thread pool is marked for termination");
+                    return std::future<typename std::result_of<F(Args...)>::type>();
+                }
+
+                std::lock_guard<std::mutex> lock(mutex_);
+
+                // Create a function with bounded parameters ready to execute
+                std::function<decltype(f(args...))()> func = std::bind(std::forward<F>(f), 
+                                                                       std::forward<Args>(args)...);
+
+                // Encapsulate it into a shared ptr in order to be able to copy construct / assign 
+                auto task_ptr = std::make_shared<std::packaged_task<decltype(f(args...))()>>(func);
+
+                // Wrap packaged task into void function
+                std::function<void()> wrapper_func = [task_ptr]() {
+                    (*task_ptr)(); 
+                };
+
+                if (true == queue_.isFull()) {
+                    spdlog::error("queue is full");
+                    return std::future<typename std::result_of<F(Args...)>::type>();
+                }
+
+                if (false == queue_.push(wrapper_func)) {
+                    spdlog::error("failed to push to queue");
+                    return std::future<typename std::result_of<F(Args...)>::type>();
+                }
+
+                /* cleanup finished threads */
+                for (size_t i = 0; i < threads_.size();) {
+                    std::future_status status = threads_[i].wait_for(0ms);
+                    if (std::future_status::ready == status) {
+                        // Remove the thread from the vector
+                        threads_.erase(threads_.begin() + i);
+                        // Decrease the count of threads
+                        numOfThreads_.fetch_sub(1);
+                    } else {
+                        ++i;
+                    }
+                }
+
+                if (numOfThreads_.load() < maxNumOfThreads_) {
+                    threads_.push_back(std::async(std::launch::async, worker, std::ref(queue_), std::ref(terminate_)));
+                    numOfThreads_.fetch_add(1);
+                }
+
+                // Return future from promise
+                return task_ptr->get_future();
+            }
+
+    private:
+
+        CyclicQueue<std::function<void()>> queue_;
+       
+        bool terminate_;
+        
+        size_t maxNumOfThreads_;
+
+        std::atomic<std::size_t> numOfThreads_;
+    
+        std::vector<std::future<void>> threads_;
+        
+        std::mutex mutex_;
+
+        static constexpr auto timeout_ = 100;
```

Comment:
Appending the unit in variable names would be helpful.

eg. timeoutSeconds_ or timeoutInSeconds_ or timeoutMilliseconds_